connectors/sources/google_drive.py

```diff
@@ -0,0 +1,638 @@
+
+            folder["path"] = "/".join(path)  # Join path elements with '/'
+
+        return folders
+
+    async def _download_content(self, blob, download_func):
+        """Downloads the file from Google Drive and returns the encoded file content.
+
+        Args:
+            blob (dict): Formatted blob document.
+            download_func (partial func): Partial function that gets the file content from Google Drive API.
+
+        Returns:
+            attachment, blob_size (tuple): base64 encoded contnet of the file and size in bytes of the attachment
+        """
+
+        temp_file_name = ""
+        blob_name = blob["name"]
+
+        logger.debug(f"Downloading {blob_name}")
+
+        async with NamedTemporaryFile(mode="wb", delete=False) as async_buffer:
+            await anext(
+                download_func(
+                    pipe_to=async_buffer,
+                )
+            )
+            temp_file_name = async_buffer.name
+
+        await asyncio.to_thread(
+            convert_to_b64,
+            source=temp_file_name,
+        )
+
+        blob_size = os.stat(temp_file_name).st_size
+
+        async with aiofiles.open(file=temp_file_name, mode="r") as target_file:
+            attachment = (await target_file.read()).strip()
+
+        logger.debug(f"Downloaded {blob_name} for {blob_size} bytes ")
+
+        await remove(str(temp_file_name))
+
+        return attachment, blob_size
+
+    async def get_google_workspace_content(self, blob, timestamp=None):
+        """Exports Google Workspace documents to an allowed file type and extracts its text content.
+
+        Shared Google Workspace documents are different than regular files. When shared from
+        a different account they don't count against the user storage quota and therefore have size 0.
+        They need to be exported to a supported file type before the content extraction phase.
+
+        Args:
+            blob (dict): Formatted blob document.
+            timestamp (timestamp, optional): Timestamp of blob last modified. Defaults to None.
+
+        Returns:
+            dict: Content document with id, timestamp & text
+        """
+
+        blob_name, blob_id, blob_mime_type = blob["name"], blob["id"], blob["mime_type"]
+
+        document = {
+            "_id": blob_id,
+            "_timestamp": blob["_timestamp"],
+        }
+
+        attachment, blob_size = await self._download_content(
+            blob=blob,
+            download_func=partial(
+                self._google_drive_client.api_call,
+                resource="files",
+                method="export",
+                fileId=blob_id,
+                mimeType=GOOGLE_MIME_TYPES_MAPPING[blob_mime_type],
+            ),
+        )
+
+        if blob_size > DEFAULT_FILE_SIZE_LIMIT:
+            logger.warning(
+                f"File size {blob_size} of file {blob_name} is larger than {DEFAULT_FILE_SIZE_LIMIT} bytes. Discarding the file content"
+            )
+            return
```

Comment:
Is it possible to not even download the file if it's too big?

Reply:
Good point, thanks for the explanation! Can you add a code comment about it?

Reply:
Also worth mentioning is that Google Export endpoint will return `403` codes for exports larger than >5MB (that's also good in preventing large downloads), see https://github.com/elastic/enterprise-search-team/issues/1374